### DISCUSSION TITLE
LLM

### DISCUSSION THREAD

**Discussion by SpudGunMan** [SpudGunMan/meshing-around#91] [Q&A]
Do you use LLM? How so? Ollama? OpenWebAI? OpenAI? 

Do you use the RAG dev code in this project?

**Comment by g7kse** [comment]
Having a go with the LLM and my poor little RPi 4 4G. Its having a hard time with it and responses are around 90 seconds. But, I'm not in a rush to go and upgrade just yet. Probably at some point to the 16G RPi.

The LLM works pretty well. It's probably not an obvious choice so perhaps it needs a bit more of an intro. A wiki maybe

edit...I think it might have killed the Pi. Are the lighter versions of the LLM better suited (TinyLlama for instance). I've no real experience with these tools

**Comment by SpudGunMan** [comment]
I have had success with Pi5/8GB and the default model at 2B parameters. under 40 seconds return. 
without a GPU and using microSD storage the results on a PI4 4GB could be quite slow yes..

**Reply by g7kse** [reply]
glacial...I'd be interested to see if [these things](https://thepihut.com/products/raspberry-pi-ai-hat-plus?variant=53621840150913) make a significant difference. I imagine they are really what is needed

**Reply by SpudGunMan** [reply]
Problem is CPU ram is slow and there is no ability to load the model up. 

I don't know ollama supports these types of cores yet due to ram problem they are good for video or voice

**Comment by PiHiker** [comment]
I use Ollama with model llama3:8b (4.7.GB) on another computer on my network with a Nvidia P5000 Quadro 16GB,  it works really well, emoji work which are neat.  I get a response back in under 10 seconds.   The history doesn't seem to work even though enabled.  Anyone else having issues with history?  History would be great to remember previous conversations it had with the node, save in file or database?  AI simulated text based games would great to try out.

**Reply by SpudGunMan** [reply]
The issue is the more history the more compute is needed to process, it considers history very very last for context and not all models consider the method of prompt. That said, there is some contextual awareness. Like I just tested with deepseek (I'm testing it out currently) and it for sure remembered and I know I can use this trick on other models such as llama

![image](https://github.com/user-attachments/assets/100fcd4b-8a91-4840-b984-221e23bbaf77)

**Reply by SpudGunMan** [reply]
also 🥧 not 🎂

**Comment by PiHiker** [comment]
Ok did some more testing! This looks good here, still using llama3:8b.


![SmartSelect_20250205_225854_Meshtastic](https://github.com/user-attachments/assets/e89f07ab-a972-4d11-8378-c5ae11168218)

It does get confused when it asks you a question back and you say "Yes" it typically looks up info on the band YES. :)

**Reply by SpudGunMan** [reply]
Very much so, check out the modules/llm.py the first few lines of code show the prompt 

The "weight" of Google has more emphasis, vs. history in the way it's currently structured with default settings.
